### PR TITLE
Reorganize Src Directory for Isomorphic React Package

### PR DESCRIPTION
**Pull request by sebmarkbage**
The new folder structure is organized around major packages that are expected to ship separately in some form.

`/isomorphic`

I moved classic/modern and children utils into a directory called "isomorphic" with the main export being ReactIsomorphic. This will eventually become the "react" package.

This includes all the dependencies that you might need to create a component without dependencies on the renderer/reconciler.

The rest moves into decoupled renderers.

`/renderers/dom/client` - This is the main renderer for DOM.

`/renderers/dom/server` - This is the server-side renderer for HTML strings.

`/addons` and `/test` - Same as before for now.

You're not supposed to take on a dependency inside another package.

Shared code is organized into a "shared" directory which is intended to support all the packages in that subdirectory. Meaning that once we swap to CommonJS modules, the only time you should use `..` is to target `../shared/` or `../../shared`.

E.g. `/shared/` is common utils that are used by everything.

`/renderers/shared/` is code that is shared by all renderers, such as the main reconciliation algorithm.

Shared code will likely be copied into each package rather than referenced. This allow us to have separate state and allow inlining and deadcode elimination.

### CODE REVIEW

**Comment by sebmarkbage** [comment]
cc @zpao

**Comment by sebmarkbage** [comment]
For my own notes. There are some unfortunate dependencies from /isomorphic right now that we should resolve.

ReactCurrentOwner.js - Needed for refs. We'll need to require this from each renderer so that they can update it.

flattenChildren.js and traverseAllChildren.js - Used both in reconcilers and children helpers. Not sure how to best share this code.

PooledClass.js - We can probably just simplify this code and drop the dependency.

ReactErrorUtils.js - This call is in the wrong place anyway which leads to bad production bugs at FB. Should move it into the event system.

ReactFragment.js - Should ideally be deprecated ASAP.

ReactUpdateQueue.js, ReactInstanceMap.js and ReactLifeCycle.js - Current the classes have a direct connection into this code. Effectively we need a pointer from the instance to the "active" renderer to get to these. Makes it not possible to make ReactInstanceMap just a Map.

ReactNativeComponent.js - Don't remember why we needed this but if we want isomorphic components we can't know ahead of time what a string element will render into. We can make heuristics at best.

**Comment by sebmarkbage** [comment]
The new folder structure is best reviewed at: https://github.com/sebmarkbage/react/tree/coreapi/src

**Comment by sebmarkbage** [comment]
Fixed lint. Now passing travis.

**Comment by zpao** [comment]
lgtm. Still not wild about "isomorphic" but it doesn't matter (yet)

**Comment by iamdustan** [comment]
Will this be getting another step closer to react-native not needing to fork react due to some of the DOM dependencies?

**Comment by iamdustan** [comment]
(asking blindly without looking at the code) :)

**Comment by sebmarkbage** [comment]
yes. Also lets server rendering use a different implementation.

File: src/renderers/dom/shared/ReactDefaultInjection.js [hunk 1]
```diff
@@ -66,7 +66,17 @@ function autoGenerateWrapperClass(type) {
...
+
 function inject() {
+  if (alreadyInjected) {
+    // TODO: This is currently true because these injections are shared between
+    // the client and the server package. They should be built independently
+    // and not share any injection state. Then this problem will be solved.
+    return;
+  }
+  alreadyInjected = true;
+
```
Comment (iamdustan) [inline src/renderers/dom/shared/ReactDefaultInjection.js]: :clap: pretty sure this is exactly what I need right now.

**Pull request merged**